packages/xrpl/test/connection.test.ts

```diff
@@ -68,250 +136,366 @@ describe('Connection', function () {
-        send(): void {
-          /* purposefully empty */
-        },
-      }
-      connection.request(mockedRequestData)
-      connection.onMessage(mockedResponse)
-      assert.deepEqual(messages, expectedMessages)
-    })
+        }
+        const connection = new Connection('url', { trace: true })
+        // @ts-expect-error -- Accessing private methods for test
+        connection.ws = {
+          send(): void {
+            /* purposefully empty */
+          },
+        }
+        const requestPromise = connection.request(mockedRequestData, 10)
+        // @ts-expect-error -- Accessing private methods for test
+        connection.onMessage(mockedResponse)
+        assert.deepEqual(messages, expectedMessages)
+        await requestPromise.catch(() => {
+          // ignore error, we intetionally fail the promise
+        })
+      },
+      TIMEOUT,
+    )
+
+    it(
+      'as a function',
+      async () => {
+        const messages: Array<[number | string, string]> = []
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any -- Need to access private methods
+        const connection: any = new Connection('url', {
+          trace(id: number | string, message: string): void {
+            messages.push([id, message])
+          },
+        })
+        connection.ws = {
+          send(): void {
+            /* purposefully empty */
+          },
+        }
+        const requestPromise = connection.request(mockedRequestData, 10)
+        connection.onMessage(mockedResponse)
+        assert.deepEqual(messages, expectedMessages)
+        await requestPromise.catch(() => {
+          // ignore error, we intetionally fail the promise
+        })
+      },
+      TIMEOUT,
+    )
   })
 
-  it('with proxy', function (done) {
-    if (isBrowser) {
-      done()
-      return
-    }
-    createServer().then((server: net.Server) => {
+  it(
+    'with proxy',
+    async () => {
+      if (isBrowser) {
+        return
+      }
+      const server = await createServer()
       const port = (server.address() as net.AddressInfo).port
       const options = {
         proxy: `ws://127.0.0.1:${port}`,
         authorization: 'authorization',
         trustedCertificates: ['path/to/pem'],
       }
-      const connection = new Connection(this.client.connection.url, options)
+      const connection = new Connection(
+        // @ts-expect-error -- Testing private member
+        clientContext.client.connection.url,
+        options,
+      )
       const expect = 'CONNECT localhost'
 
-      server.on('connection', (socket) => {
-        socket.on('data', (data) => {
-          const got = data.toString('ascii', 0, expect.length)
-          assert.strictEqual(got, expect)
-          server.close()
-          connection.disconnect()
-          done()
+      const connectionPromise = new Promise<void>((resolve) => {
+        server.on('error', () => {
+          destroyServer(server).then(() => {
+            resolve()
+          })
```

Comment:
If we destroy the server on `'error'` does the test still fail?

Reply:
I don't think so. I believe we are expecting it to fail and destroyServer is cleaning up the sockets.